### ISSUE TITLE
Intermittent "Could not establish a secure connection to the server" when using 'knife upload' against managed chef

### ISSUE THREAD

**Issue by Sam-Martin** [chef/chef#4430]
### Version:
`12.5.1` from ChefDK  `0.10.0`

### Environment:
Windows 8.1, accessing organisations on `manage.chef.io`

### Scenario:
Trying to bulk upload cookbooks using `knife upload cookbooks`

### Steps to Reproduce:

1. Create new organisation from manage.chef.io and download starter kit  
1. Download `chef_handler`, `chef-sugar`, `iis`, and `windows` cookbooks from the supermarket and extract to `cookbooks` in the chef repo (doubtful it's specific to these cookbooks, but this was the test I executed)
1. Open Windows PowerShell cd to the chef repo folder and execute `knife upload cookbooks`
1. If the error does not appear on the first run, execute `knife cookbook bulk delete .*` and try again

### Expected Result:
Upload all cookbooks from cookbooks folder to org hosted in manage.chef.io

#### Short output:
``` 
ERROR: Could not establish a secure connection to the server.
Use `knife ssl check` to troubleshoot your SSL configuration.
If your Chef Server uses a self-signed certificate, you can use
`knife ssl fetch` to make knife trust the server's certificates.

Original Exception: OpenSSL::SSL::SSLError: SSL_read: cert already in hash table
```

#### Verbose output
**From command:** ` knife upload cookbooks -VV`
http://pastebin.com/sbRJgzcd

### Additional info  
Neither adding `ssl_verify_mode :verify_none` to the knife.rb nor executing `knife ssl fetch` resolve the issue.

**Comment by lamont-granquist** [comment]
Do you have a proxy that sits in the middle of your connection to hosted?

**Comment by Sam-Martin** [comment]
@lamont-granquist nope, direct connection! Tried from home and work, same in both places.

**Comment by lamont-granquist** [comment]
Related-ish bug report on SO, but that suggests that the root cause was using cookbooks as a file distribution mechanism for many, many files -- which is clearly not the case here:

http://stackoverflow.com/questions/31019137/cert-already-in-hash-table-exception

**Comment by lamont-granquist** [comment]
Another SO post with intermittent problems that went away:

http://stackoverflow.com/questions/34131135/how-to-fix-ssl-read-cert-already-in-hash-table-on-cookbook-upload-all

**Comment by lamont-granquist** [comment]
There's a sidekiq bug report with the same error which seems to hint that concurrent access to the ruby openssl library might be an issue:  https://github.com/mperham/sidekiq/issues/1531

I believe we're using the correct calls to openssl to get threadsafety (at least I recall PRs to fix that) and this might just be a bug in ruby, particularly on windows if we're still using ruby-2.0

**Comment by lamont-granquist** [comment]
I suspect that unless someone tracks this down to a bug in how we're using openssl that the next step for this will come whenever we release omnibus builds with ruby-2.2/2.3 for windows.

**Comment by lamont-granquist** [comment]
We've upgraded to ruby 2.2.4 at least so may have picked up openssl fixes going from 2.0 to 2.1 to 2.2

I don't know what else to do here, so closing this.  If its still an issue (whenever we release 12.9.x) then it might be useful to re-open, but there's no useful work to identify to do here.
